doc/stdenv/cross-compilation: convert to markdown

siraben opened the pull request:
###### Motivation for this change
Part of [RFC 72](https://github.com/NixOS/rfcs/blob/eee65998c1d35c06906d97b45e883a1a33504cb7/rfcs/0072-commonmark-docs.md). Also, I wanted to do more substantial changes (in a future PR) to the cross-compilation docs but docbook is such a big barrier.

How to test:
```ShellSession
$ cd docs
$ nix-build && nix-shell -p python3 --run 'python3 -m http.server --directory ./result/share/doc/nixpkgs'
```

Ping @ryantm

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

siraben committed:
doc/stdenv/cross-compilation: convert to markdown

FRidh commented:
> Also, I wanted to do more substantial changes to the cross-compilation docs but docbook is such a big barrier.

Please do changes of content separately.

siraben commented:
> > Also, I wanted to do more substantial changes to the cross-compilation docs but docbook is such a big barrier.
> 
> Please do changes of content separately.

Yes, this PR just performs the conversion and some style fixes.

ryantm commented on doc/stdenv/cross-compilation.chapter.md:
```diff
@@ -0,0 +1,167 @@
+
+`hostPlatform` The "host platform" is the platform on which a package will be run. This is the simplest platform to understand, but also the one with the worst name.
+
+`targetPlatform` The "target platform" attribute is, unlike the other two attributes, not actually fundamental to the process of building software. Instead, it is only relevant for compatibility with building certain specific compilers and build tools. It can be safely ignored for all other packages.
```
I think anything in this section that was a dd/dt should be a list and have a colon and maybe also be bold. Like:

```suggestion
- *`targetPlatform`*: The "target platform" attribute is, unlike the other two attributes, not actually fundamental to the process of building software. Instead, it is only relevant for compatibility with building certain specific compilers and build tools. It can be safely ignored for all other packages.
```

jtojnar replied:
Pandoc also supports definition lists: https://pandoc.org/MANUAL.html#definition-lists

ryantm commented on doc/stdenv/cross-compilation.chapter.md:
```diff
@@ -0,0 +1,167 @@
+
+`is*` These predicates are defined in `lib.systems.inspect`, and slapped onto every platform. They are superior to the ones in `stdenv` as they force the user to be explicit about which platform they are inspecting. Please use these instead of those.
+
+`platform` This is, quite frankly, a dumping ground of ad-hoc settings (it's an attribute set). See `lib.systems.platforms` for examples—there's hopefully one in there that will work verbatim for each platform that is working. Please help us triage these flags and give them better homes!
```
Another list for these ones.

ryantm commented on doc/stdenv/cross-compilation.chapter.md:
```diff
@@ -0,0 +1,167 @@
+
+**A:** Many packages assume that an unprefixed `ar` is available, but Nix doesn't provide one. It only provides a prefixed one, just as it only does for all the other binutils programs. It may be necessary to patch the package to fix the build system to use a prefixed `ar`.
+
+**Q:** My package's testsuite needs to run host platform code.
```
This last question is missing an answer.

ryantm replied:
It's silly that some of the questions aren't actually questions, but we can fix that with later content revisions.

ryantm commented on doc/stdenv/cross-compilation.chapter.md:
````diff
@@ -0,0 +1,167 @@
+$ nix-build '<nixpkgs>' --arg crossSystem '(import <nixpkgs/lib>).systems.examples.fooBarBaz' -A whatever
+```
+
+> **Note**: Eventually we would like to make these platform examples an unnecessary convenience so that
````
This note used extend down below the shell session and paragraph.

![image](https://user-images.githubusercontent.com/4804/100965313-51396780-34df-11eb-848f-fd5d317fe8c9.png)

ryantm commented on doc/stdenv/cross-compilation.chapter.md:
```diff
@@ -0,0 +1,167 @@
+On can then imagine any sequence of platforms such that there are bootstrap stages with their 3 platforms determined by "sliding a window" that is the 3 tuple through the sequence. This was the original model for bootstrapping. Without a target platform (assume a better world where all compilers are multi-target and all standard libraries are built in their own derivation), this is sufficient. Co ...
+
+
+It is much better to refer to `buildPackages` than `targetPackages`, or more broadly package sets that do not mention "target". There are three reasons for this.
```
This used to be a note section extending down to the next note.

ryantm reviewed (changes requested):
I did a first pass. Since this is such a big section, I think it will require a few passes before we can get it dialed in. Thank you for tacking such a big and challenging section @siraben

jtojnar commented on doc/stdenv/cross-compilation.chapter.md:
```diff
@@ -0,0 +1,167 @@
+
+### Implementation of dependencies {#ssec-cross-dependency-implementation}
+
+The categorizes of dependencies developed in [Theory of dependency categorization](#ssec-cross-dependency-categorization) are specified as lists of derivations given to `mkDerivation`, as documented in [???](#ssec-stdenv-dependencies). In short, each list of dependencies for "host → target" of "foo → bar" is called `depsFooBar`, with exceptions for backwards compatibility that `depsBuildHost` is  ...
```
We might want to still use `xref`s here to avoid broken links.

jtojnar commented on doc/stdenv/cross-compilation.chapter.md:
```diff
@@ -0,0 +1,167 @@
+
+Some frequently encountered problems when packaging for cross-compilation should be answered here. Ideally, the information above is exhaustive, so this section cannot provide any new information, but it is ludicrous and cruel to expect everyone to spend effort working through the interaction of many features just to figure out the same answer to the same common problem. Feel free to add to this  ...
+
+**Q:** What if my package's build system needs to build a C program to be run under the build environment?
```
Anchors seems to have gotten lost. Either convert the questions to headings or add in-line anchors using [bracketed spans syntax](https://pandoc.org/MANUAL.html#extension-bracketed_spans) (`[]{#foo-bar}`).

ryantm commented on doc/stdenv/cross-compilation.chapter.md:
```diff
@@ -0,0 +1,185 @@
+
+: The "target platform" attribute is, unlike the other two attributes, not actually fundamental to the process of building software. Instead, it is only relevant for compatibility with building certain specific compilers and build tools. It can be safely ignored for all other packages.
+
+The build process of certain compilers is written in such a way that the compiler resulting from a single build can itself only produce binaries for a single platform. The task of specifying this single "target platform" is thus pushed to build time of the compiler. The root cause of this is that the compiler (which will be run on the host) and the standard library/runtime (which will be run on t ...
+
+There is no fundamental need to think about a single target ahead of time like this. If the tool supports modular or pluggable backends, both the need to specify the target at build time and the constraint of having only a single target disappear. An example of such a tool is LLVM.
+
+Although the existence of a "target platfom" is arguably a historical mistake, it is a common one: examples of tools that suffer from it are GCC, Binutils, GHC and Autoconf. Nixpkgs tries to avoid sharing in the mistake where possible. Still, because the concept of a target platform is so ingrained, it is best to support it as is.
```
should be indented under targetPlatform

ryantm commented on doc/stdenv/cross-compilation.chapter.md:
```diff
@@ -0,0 +1,185 @@
+
+### Theory of dependency categorization {#ssec-cross-dependency-categorization}
+
+> **Note**: This is a rather philosophical description that isn't very Nixpkgs-specific. For an overview of all the relevant attributes given to `mkDerivation`, see [Specifying dependencies](#ssec-stdenv-dependencies). For a description of how everything is implemented, see [Implementation of dependencies](#ssec-cross-dependency-implementation).
```
I think @jtojnar has admonitions working in https://github.com/NixOS/nixpkgs/pull/105284 so we should probably try to use that here too.

We can let you know when that is merged and we can rebase this on master and continue.

ryantm replied:
@siraben Admonition support has bee merged to master, so you can rebase and then use it.

ryantm commented on doc/stdenv/cross-compilation.chapter.md:
````diff
@@ -0,0 +1,185 @@
+
+In Nixpkgs, these three platforms are defined as attribute sets under the names `buildPlatform`, `hostPlatform`, and `targetPlatform`. They are always defined as attributes in the standard environment. That means one can access them like:
+
+```
````
```suggestion
```nix
```

ryantm commented on doc/stdenv/cross-compilation.chapter.md:
````diff
@@ -0,0 +1,185 @@
+
+#### What if my package's build system needs to build a C program to be run under the build environment? {#cross-qa-build-c-program-in-build-environment}
+Add the following to your `mkDerivation` invocation.
+```
````
```suggestion
```nix
```

ryantm commented on doc/stdenv/cross-compilation.chapter.md:
````diff
@@ -0,0 +1,185 @@
+####  My package's testsuite needs to run host platform code. {#cross-testsuite-runs-host-code}
+
+Add the following to your `mkDerivation` invocation.
+```
````
```suggestion
```nix
```

ryantm commented on doc/stdenv/cross-compilation.chapter.md:
```diff
@@ -0,0 +1,185 @@
+
+> **Note**: Eventually we would like to make these platform examples an unnecessary convenience so that
+>
+>     $ nix-build '<nixpkgs>' --arg crossSystem '{ config = "<arch>-<os>-<vendor>-<abi>"; }' -A whatever
```
Are we able to use backticks here so we can say it is highlighted with nix?

siraben replied:
I tried doing that but it was parsed as code, so the entire block was displayed in indented monospace.

jtojnar replied:
I think the syntax would have to be

```markdown
> **Note**: foo
>
> ```
> $ bar
> ```
```

But it looks like it still adds spaces:

https://babelmark.github.io/?text=%3E+**Note**%3A+foo%0A%3E%0A%3E+%60%60%60%0A%3E+%24+bar%0A%3E+%60%60%60

(or maybe it is a babelmark quirk, since the spaces are only shown in the rendered HTML, not in the generated HTML code :woman_shrugging:)

Or you can wait for proper admonition support from the GNOME PR:

~~~markdown
::: note
foo

```
$ bar
```
:::
~~~

siraben replied:
Ok, we should wait for the GNOME PR then I'll rebase on master.

ryantm reviewed (changes requested):
Thank you for your work and patience here. We're going to keep you in the loop about admonitions since this section has so many.

siraben commented:
Thank you for your continual feedback, I've rebased on master and changed all the note sections to use the latest admonitions.

Additionally I ran a spellchecker and fixed some spelling mistakes: platfom -> platform, langauges -> languages, bootstraping -> bootstrapping, providess -> provides.

jtojnar commented:
Looks good. Thanks.
